File: app/src/main/java/com/semisonfire/cloudgallery/utils/ExternalFileProvider.java
package com.semisonfire.cloudgallery.utils;

import android.content.Context;
import android.net.Uri;
import android.os.Environment;
import android.support.v4.content.FileProvider;

import com.semisonfire.cloudgallery.BuildConfig;

import java.io.File;

public class ExternalFileProvider extends FileProvider {

    private Context mContext;
    private File mPrivateDirectory;
    private File mPublicDirectory;

    public ExternalFileProvider() {
        super();
    }

    public ExternalFileProvider(Context context) {
        super();
        mContext = context;
    }

    public File getFile(Uri uri) {
        return new File(getPrivateDirectory(), uri.getLastPathSegment());
    }

    public Uri getUri(File file) {
        return getUriForFile(mContext, BuildConfig.APPLICATION_ID + ".provider", file);
    }

    public void setPrivateDirectory(File directory, String name) {
        mPrivateDirectory = new File(directory, name);
    }

    public File getPrivateDirectory() {
        return mPrivateDirectory;
    }

    public File getPublicDirectory() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);
    }
}

File: app/src/main/java/com/semisonfire/cloudgallery/utils/ExternalFileProvider.kt
package com.semisonfire.cloudgallery.utils

import android.net.Uri
import android.os.Environment
import android.support.v4.content.FileProvider
import com.semisonfire.cloudgallery.App
import com.semisonfire.cloudgallery.BuildConfig
import java.io.File

class ExternalFileProvider : FileProvider() {

    lateinit var privateDirectory: File
        private set

    val publicDirectory: File
        get() = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)

    fun getFile(uri: Uri): File {
        return File(privateDirectory, uri.lastPathSegment ?: "")
    }

    fun getUri(file: File): Uri {
        return getUriForFile(
            App.component.context(),
            BuildConfig.APPLICATION_ID + ".provider",
            file
        )
    }

    fun setPrivateDirectory(directory: File, name: String) {
        privateDirectory = File(directory, name)
    }

}